unimol/unimol/utils/conf_gen_cal_metrics.py

```diff
@@ -72,98 +74,130 @@ def GetDihedral(conf, atom_idx):
     mol = copy.deepcopy(tgt_mol)
     mol = Chem.AddHs(mol)
     allconformers = AllChem.EmbedMultipleConfs(
-        mol, numConfs=num_confs, randomSeed=seed, clearConfs=True
+        mol, numConfs=num_confs, randomSeed=seed, clearConfs=True, numThreads=threads
     )
-    mol = Chem.RemoveHs(mol)
-    rotable_bonds = get_torsions(mol)
-    for i in range(len(allconformers)):
-        np.random.seed(i)
-        values = 3.1415926 * 2 * np.random.rand(len(rotable_bonds))
-        for idx in range(len(rotable_bonds)):
-            SetDihedral(mol.GetConformers()[i], rotable_bonds[idx], values[idx])
-        Chem.rdMolTransforms.CanonicalizeConformer(mol.GetConformers()[i])
-    return mol
-
 
-def single_conf_gen(tgt_mol, num_confs=1000, seed=42):
-    mol = copy.deepcopy(tgt_mol)
-    mol = Chem.AddHs(mol)
-    allconformers = AllChem.EmbedMultipleConfs(
-        mol, numConfs=num_confs, randomSeed=seed, clearConfs=True
-    )
-    sz = len(allconformers)
-    for i in range(sz):
-        try:
-            AllChem.MMFFOptimizeMolecule(mol, confId=i)
-        except:
-            continue
+    # WARNING! this might change the molecule stereochemistry
+    if randomize_angles:
+        rotable_bonds = get_torsions(mol)
+        for i in range(len(allconformers)):
+            np.random.seed(i)
+            values = 3.1415926 * 2 * np.random.rand(len(rotable_bonds))
+            for idx in range(len(rotable_bonds)):
+                SetDihedral(mol.GetConformers()[i], rotable_bonds[idx], values[idx])
+            Chem.rdMolTransforms.CanonicalizeConformer(mol.GetConformers()[i])
+
+    # Forcefield relaxation improves conformer diversity
+    if mmff:
+        sz = len(allconformers)
+        for i in range(sz):
+            try:
+                AllChem.MMFFOptimizeMolecule(mol, confId=i)
+            except:
+                continue
     mol = Chem.RemoveHs(mol)
     return mol
 
 
-def single_conf_gen_no_MMFF(tgt_mol, num_confs=1000, seed=42):
-    mol = copy.deepcopy(tgt_mol)
-    mol = Chem.AddHs(mol)
-    allconformers = AllChem.EmbedMultipleConfs(
-        mol, numConfs=num_confs, randomSeed=seed, clearConfs=True
-    )
-    mol = Chem.RemoveHs(mol)
-    return mol
+def clustering(
+    mol: Chem.Mol,
+    M: int = 1000,
+    N: int = 100,
+    mmff: bool = True,
+    randomized_angles: bool = False,
+    kmeans: bool = False,
+    seed: int = 42,
+    threads: int = 0,
+) -> List[np.ndarray]:
+    """ Creates a diverse set of conformers for a given molecule by
+    procedurally generating candidates with various rdkit methods and clustering.
+    Follows principles outlined in: https://arxiv.org/abs/2302.07061
+    - For paper reproduction, call with: M=1000, N=100, randomized_angles=True, kmeans=True
+    - For best UniMol inference: M=1300, N=10, randomized_angles=False, kmeans=False (adjust M>10 for speed)
+    - WARNING! randomized_angles = True might change the molecule stereochemistry! Ex: PDB: 2ZCR
```

Comment:
explicit warning and disabled by default

Reply:
The main reason why this happens is bc rdkit's function changes the `i,j,k,l` torsion and **rotates all points linked to `l`** whereas my function in 6+T PR rotates all points linked to `k`. This prevents a change in stereochemistry when 2 different atoms are linked to `k`